packages/tupaia-web/src/features/Map/MapOverlaysLayer/PolygonLayer.tsx

```diff
@@ -0,0 +1,148 @@
+  $shade?: string;
+}
+
+const ActivePolygon = styled(Polygon)<PolygonProps>`
+  stroke: ${POLYGON_HIGHLIGHT};
+  fill-opacity: 0;
+  fill: none;
+  pointer-events: none;
+  stroke-width: 3px;
+`;
+
+const ShadedPolygon = styled(BasePolygon)<PolygonProps>`
+  fill: ${({ $shade }) => $shade};
+  fill-opacity: 0.5;
+  &:hover {
+    fill-opacity: 0.8;
+  }
+`;
+
+const BasicPolygon = styled(BasePolygon)<PolygonProps>`
+  fill: ${POLYGON_BLUE};
+  stroke: ${POLYGON_BLUE};
+  fill-opacity: 0.04;
+  stroke-width: 1;
+  &:hover {
+    fill-opacity: 0.5;
+    stroke: ${POLYGON_HIGHLIGHT};
+    fill: ${POLYGON_HIGHLIGHT};
+  }
+`;
+
+const TransparentShadedPolygon = styled(BasePolygon)<PolygonProps>`
+  fill: ${POLYGON_BLUE};
+  fill-opacity: 0;
+  &:hover {
+    fill-opacity: 0.1;
+  }
+`;
+
+interface PolygonLayerProps {
+  measureData: MeasureData[];
+  serieses: Series[];
+}
+
+const POLYGON_COMPONENTS = {
+  basicPolygon: BasicPolygon,
+  shadedPolygon: ShadedPolygon,
+  transparentShadedPolygon: TransparentShadedPolygon,
+  activePolygon: ActivePolygon,
+};
+
+const DISPLAY_TYPES = {
+  shaded: 'shadedPolygon',
+  transparentShaded: 'transparentShadedPolygon',
+  basic: 'basicPolygon',
+  active: 'activePolygon',
+};
+
+export const PolygonLayer = ({ measureData = [], serieses = [] }: PolygonLayerProps) => {
+  const { projectCode, entityCode: activeEntityCode } = useParams();
+  const navigateToEntity = useNavigateToEntity();
+  const { selectedOverlay, isPolygonSerieses } = useMapOverlays(projectCode, activeEntityCode);
+  const polygons = measureData.filter(m => !!m.region);
+
+  function getDisplayType(measure) {
+    if (
+      isPolygonSerieses &&
+      selectedOverlay?.measureLevel?.toLowerCase() === measure?.type?.toLowerCase()
+    ) {
+      // The active entity is part of the data visual so display it as a shaded polygon rather
+      // than an active polygon
+      return DISPLAY_TYPES.shaded;
+    }
+    if (measure?.code === activeEntityCode) {
+      return DISPLAY_TYPES.active;
+    }
+    if (measure.isHidden) {
+      return DISPLAY_TYPES.basic;
+    }
+    if (isPolygonSerieses) {
+      return DISPLAY_TYPES.shaded;
+    }
+
+    return DISPLAY_TYPES.basic;
+  }
+
+  return (
+    <LayerGroup>
+      {polygons.map((measure: MeasureData) => {
+        const { region, code, color, name, permanentTooltip = false } = measure;
+        const shade = BREWER_PALETTE[color as keyof typeof BREWER_PALETTE] || color;
+        const displayType = getDisplayType(measure);
+        const PolygonComponent = POLYGON_COMPONENTS[displayType];
+        const key =
+          displayType === DISPLAY_TYPES.active ? `currentEntityPolygon${Math.random()}` : code;
+        const showDataOnTooltip = displayType === DISPLAY_TYPES.shaded;
+
+        return (
+          <PolygonComponent
+            positions={region}
+            $displayType={displayType}
+            $shade={shade}
+            eventHandlers={{
+              click: () => {
```

Comment:
Are there some cases where the polygon isn't clickable? I think at the moment there are, but maybe that's a weird thing to add in 🤔

Reply:
hmm I had a look and the only case I can think of is the active entity so I updated it to handle that.